cblox/include/cblox/core/submap_collection_inl.h

```diff
@@ -0,0 +1,376 @@
+  if (!outfile.is_open()) {
+    LOG(ERROR) << "Could not open file for writing: " << file_path;
+    return false;
+  }
+  // Saving the submap collection header object
+  TsdfSubmapCollectionProto tsdf_submap_collection_proto;
+  getProto(&tsdf_submap_collection_proto);
+  // Write out the layer header.
+  if (!voxblox::utils::writeProtoMsgToStream(tsdf_submap_collection_proto,
+                                             &outfile)) {
+    LOG(ERROR) << "Could not write submap collection header message.";
+    outfile.close();
+    return false;
+  }
+  // Saving the tsdf submaps
+  for (const auto& id_submap_pair : id_to_submap_) {
+    // DEBUG
+    std::cout << "Saving tsdf_submap with ID: " << id_submap_pair.first
+              << std::endl;
+    // Saving the submap
+    (id_submap_pair.second)->saveToStream(&outfile);
+  }
+  // Closing the file
+  outfile.close();
+  return true;
+}
+
+template <typename SubmapType>
+void SubmapCollection<SubmapType>::getProto(
+    TsdfSubmapCollectionProto* proto) const {
+  // Checks
+  CHECK_NOTNULL(proto);
+  // Filling out the description of the submap collection
+  proto->set_voxel_size(submap_config_.tsdf_voxel_size);
+  proto->set_voxels_per_side(submap_config_.tsdf_voxels_per_side);
+  proto->set_num_submaps(num_patches());
+}
+
+// Fusing the submap pairs
+template <typename SubmapType>
+void SubmapCollection<SubmapType>::fuseSubmapPair(
+    const SubmapIdPair& submap_id_pair) {
+  // Extracting the submap IDs
+  SubmapID submap_id_1 = submap_id_pair.first;
+  SubmapID submap_id_2 = submap_id_pair.second;
+  // DEBUG
+  std::cout << "Fusing submap pair: (" << submap_id_1 << ", " << submap_id_2
+            << ")" << std::endl;
+  // Getting the requested submaps
+  auto id_submap_pair_1_it = id_to_submap_.find(submap_id_1);
+  auto id_submap_pair_2_it = id_to_submap_.find(submap_id_2);
+  // If the submaps are found
+  if ((id_submap_pair_1_it != id_to_submap_.end()) &&
+      (id_submap_pair_2_it != id_to_submap_.end())) {
+    // Getting the submaps
+    typename SubmapType::Ptr submap_ptr_1 = (*id_submap_pair_1_it).second;
+    typename SubmapType::Ptr submap_ptr_2 = (*id_submap_pair_2_it).second;
+    // Checking that we're not trying to fuse a submap into itself. This can
+    // occur due to fusing submap pairs in a triangle.
+    if (submap_ptr_1->getID() == submap_ptr_2->getID()) {
+      std::cout << "Avoided fusing submap into itself." << std::endl;
+      return;
+    }
+    // Getting the tsdf submap and its pose
+    const Transformation& T_M_S1 = submap_ptr_1->getPose();
+    const Transformation& T_M_S2 = submap_ptr_2->getPose();
+    const Transformation& T_S1_S2 = T_M_S1.inverse() * T_M_S2;
+    // Merging the submap layers
+    mergeLayerAintoLayerB(submap_ptr_2->getTsdfMap().getTsdfLayer(), T_S1_S2,
+                          submap_ptr_1->getTsdfMapPtr()->getTsdfLayerPtr());
+    // Transfering KeyFrames to the new submap
```

Comment:
More MM code... I think maybe all this should be put in an archive file somwewhere to avoid rewritting it when I integrate orb_slam